### ISSUE TITLE
Assembler refactoring

### ISSUE THREAD

**Issue by melkyades** [powerlang/powerlang#56]
The initial idea we had was to add an object in between our current assembler and the method nativizer. After looking at the code, I think it would be better to just rename Assembler64/JITAssembler64 to AssemblerAMD64 and JITAssemblerAMD64, and let other assemblers to mimic its interface (well, we don't have interfaces in Smalltalk but it should be like implementing an interface). 

So let's say, AssemblerRISCV would implement most of current JITAssembler64 interface, but in the way the RISCV implementor wants. The nativization process would add an object in charge of defining the target architecture (maybe it could be ABI, maybe we need something else), and this object would be the one in charge of instantiating the corresponding assembler.

**Comment by janvrany** [comment]
This is what I thought initially but that does not fly.

Consider: 
```
JITAssembler64 >> restoreCallerEnvironment
    self loadEwithFPindex: -2   
```
```
JITAssembler64 >> loadEwithFPindex: index
    self assemble: 'mov' withReg: self regE withReg: self regFP index: index  
```
```
Assembler64 >> assemble: mnemonic
withReg: dst
withReg: src
index: index
    pointer
        reset;
        length: self addressLength;
        base: src;
        displacement: index - 1 * wordSize.
    self assemble: mnemonic with: dst with: pointer    
```

I do not want to copy `JITAssembler64 >> restoreCallerEnvironment` into (new) `AssemblerRISCV`. 

Consider 
```
incA
    self assemble: 'inc' with: self regA 
```
```
incC
    self assemble: 'inc' with: self regC 
```

I do not want to implement both for new backend when implementing just one per backend would suffice. 

Looking at `Assembler64` and `JITAssembler64`, both combined contain 478 methods, nearly 300 are used from outside:

```
classes := Smalltalk allClasses select: [:e | e category notNil and:[e category startsWith: 'Powerlang' ] ].
selectors := Assembler64 selectors , JITAssembler64 selectors.
refd := Set new.
selectors do: [:selector | 
    | senders |

    senders := SystemBrowser findSendersOf: selector
                in: classes
                ignoreCase: false.
    senders do: [:sender | 
        (sender mclass ~~ Assembler64 and: [ sender mclass ~~ JITAssembler64 ]) ifTrue: [ 
            refd add: selector.
        ].
    ].
].
refd size 
```

No me gusta!

**Comment by melkyades** [comment]
> Notice however that this doesn't mean you have to manually implement these things. If we are able to define semantics based on the selector we could automatically generate the backends by using the ISA specification.

@shingarov what do you think?

**Comment by janvrany** [comment]
> I think the higher level methods might be moved to the method nativizer or a high level object used by the assembler when needed, instead of being always called by this intermediate thing.

Yes, and this is what that all talk about 
> The initial idea we had was to add an object in between our current assembler and the method nativizer
was. 

But I'd prefer to do this *before* attempting to implement another backend. Doing it after is waste of time, because then it is just an extra work. 

So, what we need is a base class with bunch of `#subclassResponsibility` defining the backend API. These are the "tiles"
or ("evaluators" as they're called in TR or RTL instructions as they're called in Cog, IIRC). These make clear separation on what is backend-agnostic and what is backend-specific. Now, there's no such clear boundary. 

> this is unusable by other ISAs, mov semantics can be different in each architecture (does riscv even have a mov? I thought it was add, r1, 0, r2)

No, that was not my point. Clearly, you can implement semantics of `loadEwithFPindex: index` in pretty much any ISA. 
But having to re-implement JITAssembler64 API means I have to implement both, `restoreCallerEnvironment` and that `loadE...`. The problem is that got class `JITAssembler64` does too many things, part of it is in its superclass. 

What is my rant all about is that one object should use "tiles" (or IR operations) and then another implements those tiles (i.e, chose what instructions use to implement semantics of the "tiles" (or IR operations). These two must be separate and relation between them should not be inheritance, but composition. This is all I want at this point. 

> Again, these instructions don't exist in other ISAs, so you'll have to impement one per backend.
Sure, but why on earth I have to implement two methods when we can do something like: 

```
IRCompiler >> incA
   codegen inc: self regA

IRCompiler >> incC
   codegen inc: self reggC
```
Then all I need is to implement *ONE* method in `codegen`, not two. Or, not even that, it makes perfect sense have
```
AbstractCodeGenerator >> addi: immediate to: reg
  self subclassResponsibility

AbstractCodeGenerator >> inc: reg
   self addi: 1 to: reg

X86CodeGenerator >> addi: immediate to: reg
  self assemble: ...

X86CodeGenerator >> inc: reg
   "Suppose we want to use inc over add..."
   self assemble: ...
```

Then, in most backends I do not have to care about `incA` and `incC` at all, assuming that we need `addi:to:` anyways for other purposes.

> Notice however that this doesn't mean you have to manually implement these things. If we are able to define semantics 
Yes, but we're not there. Besides, now we do not have clear definition of things we would eventually need to define semantics of. Again, doing the split we agreed on a year ago and for which I'm going to fight fiercely, would give us a clear list.

So, let's do the split. "|moved to the method nativizer or a high level object" first.

**Comment by janvrany** [comment]
Speaking of RISC-V, even it one would implement `RISCVAssembler` as drop-in replacement for (what is now) `JITAssembler64`, it would not immediately fly. 

Consider:
```
BinaryMessageNativizer >> inlineMultiply
  "..."
  assembler
        pushA;
        convertAtoNativeInteger;
        loadTwithR;
        clearRintegerBit;
        multiplyRbyA;
        popA;
        jumpIfOverflowTo: overflow;
        ensureRintegerBit;                   
        ...
```
The problem is that there are architectures without condition register - RISC-V is one of them. All you have is 
"compare and branch" instructions. The method above is particularly troublesome, because in between doing
the arithmetic operation and "jumping if overflow" you clobber the arithmetic operand and on RISC-V, the only way to check overflow is to perform checks on resulting value after the arithmetic operation (see RISC-V Unprivileged ISA V20191213, page 18 if interested)So the design of "tiles" / IR where one tile / IR op is "compare" or "add" and other is 
"jump if overflow" or "jump if equal" just does not work. 
This is an example of what I meant when I said that "intel architecture is assumed" (though for sure, other architectures like POWER or AA64 have condition registers, so no big deal there)

But this is the next step, after the split. 

Interestingly enough, overflow handling is now being discussed in RISC-V J-ext group, but adding a condition register is off the table - one option discussed was to add U-mode trap, in that case, assembler refactoring is a tiny job compared to 
refactoring we'd have to do to support that. But don't worry, this is nowhere near to materialize :-)

**Comment by shingarov** [comment]
> The initial idea we had was to add an object in between our current assembler and the method nativizer.

Yes indeed; one of the problems we talked about, was the API forcing us (through hard-coding) to lump together concerns of object-representation / internal-ABI-conventions, of instruction selection, of encoding, and of spitting-at-istream.

But this is only one of the problems.  Another one is, how do you tell which of all this Smalltalk-80 code is runtime and which is compiler?  For example, Eliot's Cog contains an enormous amount of asserts in the runtime.  Right now these are used just to help the programmer's confidence somewhat, still remaining within pointwise computing; but I can't think of any obstacle why these asserts can't easily be the basis for a full correctness proof.

Now.  This is the *runtime*.  Things are quite different with the JIT.  For an oversimplified example, say I want to nativize an operation of zeroing-out the rClass register.  Its Hoare triple, i.e.

{ _preconditions_ } _computation_ { _postconditions_ }

would be something like

```
zeroOutRClass
    { }
    mov rClass, 0
    { self assert: rClass equals: 0 }
```

(Actually, we could even put the assembly code in the concrete subclasses, and the pre/postconditions in the abstract superclass; and whenever we have a superoptimizer (such as PIG) we will be able to omit the subclasses altogether.  But this is not relevant to this discussion)

The problem why (e.g.) Cog can't do something like this, is because we need what Pocho calls "time travel".  Only that in the ULD demo we "traveled-back" and here we "travel-forward".  This shouldn't be difficult.  In fact, we already have most of the "meat" of the mechanics.  Well. at least prototypes: the assembler to encode and spit at the symbolic instruction memory, has been there for a long time; a prototype for lifting those symbolic instructions, I showed back in Athens so we know how to do it; for the checking of the postconditions we still rely on Python code from Angr but it's just gruntwork "to have it at our Smalltalk fingertips" (as Eliot likes to put it).  What we *don't* have yet, is the "connecting tissue": how to express this separation of cross-cutting concerns in Smalltalk.  In the proof-of-concept embedding of the TA-Assembler within the Smalltalk compiler (a quick Opal plugin hack) I massage bytecode along the lines of what the Aspect guys are doing ( https://hal.inria.fr/inria-00470400v3/document ), but I intensely dislike that "PDLOpalPlugin".  I've been flirting with CLEAN's linear approach for a while now, but this hasn't taken any concrete shape yet.  Be it as it may, we have strayed off the topic of Nativizer API.  What *is* relevant to the topic, is that the API must at least *allow* the clear separation of "native code tile" concept, and not force to confuse it with "the code that emits it" and "the time when said code emits it" by forcing these concerns to be hardcoded as one.

**Comment by shingarov** [comment]
> What is my rant all about is that one object should use "tiles" (or IR operations) and then another implements those tiles (i.e, chose what instructions use to implement semantics of the "tiles" (or IR operations). These two must be separate and relation between them should not be inheritance, but composition. This is all I want at this point.

Yes, I agree this is the minimum starting point.  And we already wrote the skeleton structure of these classes last summer, and we implemented **one** path of calls at the Hackathon, so all that remains is fill the rest of the flesh (which can be just copy/pasted from the current Assembler).  Cog also has these classes (Cogit and RTLCompiler) plus I added a bunch of support for ULD time-travel which I should disentangle from Cog and make it generic across nativizers.

**Comment by shingarov** [comment]
>This is an example of what I meant when I said that "intel architecture is assumed" (though for sure, other architectures like POWER or AA64 have condition registers, so no big deal there)

Exactly.  And btw, it **is** a big deal on POWER because its "condition register" is not what Intel or ARM people are used to, where each flag has a fixed meaning.  On POWER the CR register is a generous (32 bits) storage for flags and you get to decide (subject to complex restraint rules varying per instruction) what bits to route in and out of there.  For example, on ARM there is both a "greater" flag and "above" flag for signed and unsigned comparison, but on POWER there is only one bit (per each of the 8 sub-registers) and its meaning will depend on the instruction.

**Comment by shingarov** [comment]
>The method above is particularly troublesome, because in between doing the arithmetic operation and "jumping if overflow" you clobber the arithmetic operand and on RISC-V, the only way to check overflow is to perform checks on resulting value after the arithmetic operation

That's a bigger problem than just refactoring the Tiler and Concretizer into separate classes.  In this case, the problem is that the tiles are too small.  So it's not just refactoring, but carefully designing the "TRIL abstract ISA" (or "RTL abstract ISA" in Cog words).

**Comment by janvrany** [comment]
> And btw, it is a big deal on POWER
I stand corrected. POWER is weird : - )

> That's a bigger problem than just refactoring the Tiler and Concretizer
Sure. But one step at time. You see, this is why I was talking about major overhaul, splitting is one step in that effort - or so I see it.

**Comment by melkyades** [comment]
> But I'd prefer to do this _before_ attempting to implement another backend. Doing it after is waste of time, because then it is just an extra work.

I could argue the other way around. I think both approaches have problems that are hard to see until you get there and either you or I will end doing extra work. I think it's best to do both at the same time so we detect problems as early as possible.

> 
> So, what we need is a base class with bunch of `#subclassResponsibility` defining the backend API. These are the "tiles"
> or ("evaluators" as they're called in TR or RTL instructions as they're called in Cog, IIRC). These make clear separation on what is backend-agnostic and what is backend-specific. Now, there's no such clear boundary.

We can make a base abstract JITAssembler class.

> 
> > this is unusable by other ISAs, mov semantics can be different in each architecture (does riscv even have a mov? I thought it was add, r1, 0, r2)
> 
> No, that was not my point. Clearly, you can implement semantics of `loadEwithFPindex: index` in pretty much any ISA.
> But having to re-implement JITAssembler64 API means I have to implement both, `restoreCallerEnvironment` and that `loadE...`. The problem is that got class `JITAssembler64` does too many things, part of it is in its superclass.
> 
> What is my rant all about is that one object should use "tiles" (or IR operations) and then another implements those tiles (i.e, chose what instructions use to implement semantics of the "tiles" (or IR operations). These two must be separate and relation between them should not be inheritance, but composition. This is all I want at this point.

Please don't rant anymore, everybody here already has their own problems and the internet is full of places where you can go and rant, not here. 

> 
> > Again, these instructions don't exist in other ISAs, so you'll have to impement one per backend.
> > Sure, but why on earth I have to implement two methods when we can do something like:
> 
> ```
> IRCompiler >> incA
>    codegen inc: self regA
> 
> IRCompiler >> incC
>    codegen inc: self reggC
> ```
> 
> Then all I need is to implement _ONE_ method in `codegen`, not two. Or, not even that, it makes perfect sense have
> 
> ```
> AbstractCodeGenerator >> addi: immediate to: reg
>   self subclassResponsibility
> 
> AbstractCodeGenerator >> inc: reg
>    self addi: 1 to: reg
> 
> X86CodeGenerator >> addi: immediate to: reg
>   self assemble: ...
> 
> X86CodeGenerator >> inc: reg
>    "Suppose we want to use inc over add..."
>    self assemble: ...
> ```
> 
> Then, in most backends I do not have to care about `incA` and `incC` at all, assuming that we need `addi:to:` anyways for other purposes.

That looks nice but it doesn't work that well. One of the reasons of having things like `incA` and `incC` on the first place is that they can be templatized. This was already a performance bottleneck in the dmr.

> 
> > Notice however that this doesn't mean you have to manually implement these things. If we are able to define semantics
> > Yes, but we're not there. Besides, now we do not have clear definition of things we would eventually need to define semantics of. Again, doing the split we agreed on a year ago and for which I'm going to fight fiercely, would give us a clear list.
> 
> So, let's do the split. "|moved to the method nativizer or a high level object" first.

I'll open a branch so that we can pair program next week.

**Comment by janvrany** [comment]
> I'll open a branch so that we can pair program next week.
Perfect!

**Comment by melkyades** [comment]
branch is now open: 95d32784566ae3be6586826e0191bcd233024b14

**Comment by janvrany** [comment]
I have just opened a draft PR: #57, mainly to allow people commenting and to have a CI. 
There's still a lot to be done.
